rust/src/ics23.rs

```diff
@@ -0,0 +1,128 @@
+extern crate failure;
+
+
+use crate::proofs;
+use crate::verify::{CommitmentRoot, verify_existence};
+
+pub fn verify_membership(proof: &proofs::CommitmentProof, spec: &proofs::ProofSpec, root: CommitmentRoot, key: &[u8], value: &[u8]) -> bool {
+    match &proof.proof {
+        Some(proofs::CommitmentProof_oneof_proof::exist(ex)) => {
+            let valid = verify_existence(&ex, spec, root, key, value);
+            valid.is_ok() && valid.unwrap()
+        }
+        _ => false,
+    }
+}
+
+pub fn iavl_spec() -> proofs::ProofSpec {
+    let mut leaf = proofs::LeafOp::new();
+    leaf.set_prefix(vec![0u8]);
+    leaf.set_hash(proofs::HashOp::SHA256);
+    leaf.set_prehash_value(proofs::HashOp::SHA256);
+    leaf.set_length(proofs::LengthOp::VAR_PROTO);
+
+    let mut spec = proofs::ProofSpec::new();
+    spec.set_leaf_spec(leaf);
+    spec
+}
+
+pub fn tendermint_spec() -> proofs::ProofSpec {
+    let mut leaf = proofs::LeafOp::new();
+    leaf.set_prefix(vec![0u8]);
+    leaf.set_hash(proofs::HashOp::SHA256);
+    leaf.set_prehash_value(proofs::HashOp::SHA256);
+    leaf.set_length(proofs::LengthOp::VAR_PROTO);
+
+    let mut spec = proofs::ProofSpec::new();
+    spec.set_leaf_spec(leaf);
+    spec
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    extern crate protobuf;
+    extern crate serde;
+    extern crate serde_json;
```

Comment:
in general, `extern crate` is not necessary when using rust 2018, with a [couple rare exceptions](https://doc.rust-lang.org/edition-guide/rust-2018/module-system/path-clarity.html#an-exception)